### PR TITLE
refactor(profiling/memalloc): add array handling macros

### PR DESCRIPTION
**Pull request by jd**
This refactors the current traceback list into array where utility functions
are automatically generated from macros.

### CODE REVIEW

File: ddtrace/profiling/collector/_utils.h [hunk 1]
```diff
@@ -0,0 +1,116 @@
+#ifndef _DDTRACE_UTILS_H
+#define _DDTRACE_UTILS_H
+
+#include <Python.h>
+#include <stdlib.h>
+
+#define p_new(type, count) PyMem_RawMalloc(sizeof(type) * (count))
+#define p_delete(mem_p) PyMem_Free(mem_p);
+#define p_alloc_nr(x) (((x) + 16) * 3 / 2)
```
Comment (Kyle-Verhoog) [inline ddtrace/profiling/collector/_utils.h:9]: A comment here might be useful. These seem like magic numbers

Comment (jd) [reply ddtrace/profiling/collector/_utils.h:9]: Added.

File: ddtrace/profiling/collector/_utils.h [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+/** Common array type */
+#define ARRAY_TYPE(type_t, pfx, size_type)                                                                             \
+    typedef struct pfx##_array_t                                                                                       \
+    {                                                                                                                  \
+        type_t* tab;                                                                                                   \
+        size_type count, size;                                                                                         \
+    } pfx##_array_t;
+
+#define foreach(var, array)                                                                                            \
```
[the remark this answers is not in the document]
Comment (jd) [reply ddtrace/profiling/collector/_utils.h:35]: There are going to be used in the next PR :)

File: ddtrace/profiling/collector/_utils.h [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    for (typeof((array).count) __foreach_index_##var = 0; __foreach_index_##var < (array).count;                       \
+         __foreach_index_##var = (array).count)                                                                        \
```
[the remark this answers is not in the document]
Comment (jd) [reply ddtrace/profiling/collector/_utils.h:37]: Here:
we start at `foreach_index` = 0
then go until `foreach_index` goes above count (that's true at first iteration)
after the first iteration, it sets `foreach_index` to `count` to make the condition false and the loop stops right away.

This means this `for` loop only is run once but `foreach_index` has already the correct value `0`.

File: ddtrace/profiling/collector/_utils.h [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        for (typeof((array).tab) var = &(array).tab[__foreach_index_##var];                                            \
+             (__foreach_index_##var < (array).count) && (var = &(array).tab[__foreach_index_##var]);                   \
```
[the remark this answers is not in the document]
Comment (jd) [reply ddtrace/profiling/collector/_utils.h:39]: Here:
`var` is the first item of the array at init
then the loop runs while`foreach_index` is below the `count` of items and assign `var` to that item.
when done, increment `foreach_index` by 1.

The use of 2 for loops is a trick to be able to init 2 variables, `foreach_index` and `var`.

This is a little painful to write, but it's then optimized down by the compiler and this syntax makes sure `break` works correctly. 😁

File: ddtrace/profiling/collector/_utils.h [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+#define foreach_reverse(var, array)                                                                                    \
+    for (typeof((array).count) __foreach_index_##var = 0; __foreach_index_##var < (array).count;                       \
+         __foreach_index_##var = (array).count)                                                                        \
+        for (typeof((array).tab) var = &(array).tab[(array).count - __foreach_index_##var];                            \
+             (__foreach_index_##var < (array).count) && (var = &(array).tab[__foreach_index_##var]);                   \
+             ++__foreach_index_##var)
+
+/** Common array functions */
+#define ARRAY_COMMON_FUNCS(type_t, pfx, size_type, dtor)                                                               \
```
[the remark this answers is not in the document]
Comment (jd) [reply ddtrace/profiling/collector/_utils.h:50]: This was done so it's possible to have a different set of functions defined for an array.
For example, you could have an array being a set or being always-ordered, and that'd need a different type of function, meaning you'd have those "common" ones plus others.

File: ddtrace/profiling/collector/_utils.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (Kyle-Verhoog) [reply ddtrace/profiling/collector/_utils.h]: Ah ok, I think I got it now lol

**Comment by jd** [comment]
I've removed the `foreach` macros for now :)

**Review by Kyle-Verhoog — APPROVED** [review]

**Pull request merged**